Components/FlightControl/RocketSM.cpp

```diff
@@ -0,0 +1,455 @@
+/**
+ ******************************************************************************
+ * File Name          : RocketSM.cpp
+ * Description        : Primary rocket state machine
+ ******************************************************************************
+*/
+#include "RocketSM.hpp"
+#include "SystemDefines.hpp"
+/* Rocket Control ------------------------------------------------------------------*/
+/**
+ * @brief Default constructor for Rocket SM, initializes all states
+ */
+RocketControl::RocketControl()
+{
+    // Setup the internal array of states. Setup in order of enum.
+    stateArray[RS_PRELAUNCH] = new PreLaunch();
+    stateArray[RS_FILL] = new Fill();
+    stateArray[RS_ARM] = new Arm();
+    stateArray[RS_IGNITION] = new Ignition();
+    stateArray[RS_LAUNCH] = new Launch();
+    stateArray[RS_BURN] = new Burn();
+    stateArray[RS_COAST] = new Coast();
+    stateArray[RS_DESCENT] = new Descent();
+    stateArray[RS_RECOVERY] = new Recovery();
+    stateArray[RS_ABORT] = new Abort();
+
+    // Verify all states are initialized
+    for (uint8_t i = 0; i < RS_NONE-1; i++)
+        SOAR_ASSERT(stateArray[i] != nullptr);
+}
+
+/**
+ * @brief Handles state transitions
+ * @param nextState The next state to transition to
+ * @return The state after the transition
+ */
+RocketState RocketControl::TransitionState(RocketState nextState)
+{
+    // Check if we're already in the next state (TransitionState does not allow entry into the existing state)
+    if (nextState == rs_currentState->GetStateID())
+        return rs_currentState->GetStateID();
+
+    // Check the next state is valid
+    if (nextState >= RS_NONE)
+        return rs_currentState->GetStateID();
```

Comment:
also need to check if a transition to nextState is possible from rs_currentState (ie. this is where we program hard locks on transitions, for example, from fill to ignition)

Reply:
Hard locks on transitions are implemented by the fact that a given state's HandleCommand function does not implement the functionality to transition to a given state. It should not be necessary to implement a double lock here.